File: deepseek_balance.py
#!/usr/bin/env python3
"""
DeepSeek 余额 macOS 菜单栏应用
在状态栏实时显示 DeepSeek API 余额
"""

import os
import json
import sqlite3
from datetime import datetime, timedelta
from pathlib import Path

import requests
import rumps
from AppKit import NSApplication, NSApplicationActivationPolicyAccessory

# ============================================================
# 路径配置
# ============================================================
APP_DIR = Path(os.path.dirname(os.path.abspath(__file__)))
CONFIG_PATH = APP_DIR / "config.json"
DATA_DIR = APP_DIR / "data"
DB_PATH = DATA_DIR / "balance_history.db"

DATA_DIR.mkdir(parents=True, exist_ok=True)


# ============================================================
# 配置管理
# ============================================================
def load_config():
    with open(CONFIG_PATH, 'r') as f:
        return json.load(f)


def save_config(config):
    with open(CONFIG_PATH, 'w') as f:
        json.dump(config, f, indent=4, ensure_ascii=False)


# ============================================================
# 数据库
# ============================================================
def init_db():
    conn = sqlite3.connect(str(DB_PATH))
    cursor = conn.cursor()
    cursor.execute('''
        CREATE TABLE IF NOT EXISTS balance_history (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            timestamp TEXT NOT NULL,
            total_balance REAL NOT NULL,
            granted_balance REAL NOT NULL,
            topped_up_balance REAL NOT NULL,
            is_available INTEGER NOT NULL
        )
    ''')
    conn.commit()
    conn.close()


def save_balance_record(total, granted, topped, is_available):
    conn = sqlite3.connect(str(DB_PATH))
    cursor = conn.cursor()
    now = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    cursor.execute('''
        INSERT INTO balance_history
        (timestamp, total_balance, granted_balance, topped_up_balance, is_available)
        VALUES (?, ?, ?, ?, ?)
    ''', (now, total, granted, topped, 1 if is_available else 0))
    conn.commit()
    conn.close()


def get_today_usage():
    """今日使用量"""
    conn = sqlite3.connect(str(DB_PATH))
    cursor = conn.cursor()
    today = datetime.now().strftime('%Y-%m-%d')
    cursor.execute('''
        SELECT total_balance FROM balance_history
        WHERE timestamp >= ? ORDER BY timestamp ASC LIMIT 1
    ''', (today + ' 00:00:00',))
    first = cursor.fetchone()
    cursor.execute('''
        SELECT total_balance FROM balance_history
        WHERE timestamp >= ? ORDER BY timestamp DESC LIMIT 1
    ''', (today + ' 00:00:00',))
    last = cursor.fetchone()
    conn.close()
    if first and last:
        return first[0] - last[0]
    return 0.0


# ============================================================
# DeepSeek API
# ============================================================
def fetch_balance(api_key, base_url):
    url = f"{base_url}/user/balance"
    headers = {
        'Accept': 'application/json',
        'Authorization': f'Bearer {api_key}'
    }
    try:
        resp = requests.get(url, headers=headers, timeout=15)
        resp.raise_for_status()
        return resp.json()
    except requests.exceptions.Timeout:
        raise Exception("请求超时，请检查网络")
    except requests.exceptions.HTTPError as e:
        if e.response.status_code == 401:
            raise Exception("API Key 无效，请在设置中更新")
        elif e.response.status_code == 429:
            raise Exception("请求过于频繁，请稍后重试")
        else:
            raise Exception(f"API 错误 (HTTP {e.response.status_code})")
    except requests.exceptions.ConnectionError:
        raise Exception("无法连接 DeepSeek 服务器")


# ============================================================
# 菜单栏应用
# ============================================================
class DeepSeekBalanceApp(rumps.App):
    def __init__(self):
        # 纯菜单栏应用，不显示 Dock 图标
        NSApplication.sharedApplication().setActivationPolicy_(NSApplicationActivationPolicyAccessory)

        # icon_path = str(APP_DIR / "icon.png")
        super().__init__(
            name="DeepSeekBalance",
            title="---",
            icon=str(APP_DIR / "icon.png"),
            quit_button=None
        )
        self.config = load_config()
        self.api_key = self.config.get('api_key', '')
        self.base_url = self.config.get('base_url', 'https://api.deepseek.com')
        self.poll_interval = self.config.get('poll_interval_minutes', 5) * 60

        self.current_balance = None
        self.is_loading = False

        init_db()
        self._build_menu()

        # 定时轮询
        self.timer = rumps.Timer(self._fetch_and_update, self.poll_interval)
        self.timer.start()

        # 首次立即加载
        self._fetch_and_update()

    def _build_menu(self):
        self.menu = [
            rumps.MenuItem("🔄 加载中...", callback=None),
            None,
            rumps.MenuItem("🔄 立即刷新", callback=self._force_refresh),
            rumps.MenuItem("⚙️ 设置...", callback=self._show_settings),
            None,
            rumps.MenuItem("❌ 退出", callback=self._quit_app),
        ]
        self.menu['🔄 加载中...'].set_callback(None)

    def _update_menu(self):
        if self.current_balance:
            info = self.current_balance
            is_avail = info.get('is_available', True)
            balance_list = info.get('balance_infos', [])
            if balance_list:
                b = balance_list[0]
                total = float(b.get('total_balance', 0))
                granted = float(b.get('granted_balance', 0))
                topped = float(b.get('topped_up_balance', 0))
                symbol = '¥' if b.get('currency', 'CNY') == 'CNY' else '$'

                self.title = f"{symbol}{total:.2f}"

                today_usage = get_today_usage()

                self.menu.clear()
                self.menu.add(rumps.MenuItem(
                    f"📊 总余额: {symbol}{total:.2f}", callback=None))
                self.menu.add(rumps.MenuItem(
                    f"🎁 赠送余额: {symbol}{granted:.2f}", callback=None))
                self.menu.add(rumps.MenuItem(
                    f"💳 充值余额: {symbol}{topped:.2f}", callback=None))
                self.menu.add(None)
                if today_usage > 0:
                    self.menu.add(rumps.MenuItem(
                        f"📉 今日使用: -{symbol}{today_usage:.2f}", callback=None))
                else:
                    self.menu.add(rumps.MenuItem(
                        f"📉 今日使用: {symbol}0.00", callback=None))
                self.menu.add(None)
                self.menu.add(rumps.MenuItem("🔄 立即刷新", callback=self._force_refresh))
                self.menu.add(rumps.MenuItem("⚙️ 设置...", callback=self._show_settings))
                self.menu.add(None)
                self.menu.add(rumps.MenuItem("❌ 退出", callback=self._quit_app))
                return
        self.title = "..."

    def _update_error(self, msg):
        self.title = "⚠️"
        self.menu.clear()
        self.menu.add(rumps.MenuItem(f"❌ {msg}", callback=None))
        self.menu.add(None)
        self.menu.add(rumps.MenuItem("🔄 重试", callback=self._force_refresh))
        self.menu.add(rumps.MenuItem("⚙️ 设置...", callback=self._show_settings))
        self.menu.add(None)
        self.menu.add(rumps.MenuItem("❌ 退出", callback=self._quit_app))

    def _fetch_and_update(self, _=None):
        if self.is_loading:
            return
        self.is_loading = True
        try:
            data = fetch_balance(self.api_key, self.base_url)
            self.current_balance = data
            balance_list = data.get('balance_infos', [])
            if balance_list:
                b = balance_list[0]
                save_balance_record(
                    float(b.get('total_balance', 0)),
                    float(b.get('granted_balance', 0)),
                    float(b.get('topped_up_balance', 0)),
                    data.get('is_available', True)
                )
            self._update_menu()
        except Exception as e:
            self._update_error(str(e))
        finally:
            self.is_loading = False

    def _force_refresh(self, _):
        self.title = "..."
        self.menu.clear()
        self.menu.add(rumps.MenuItem("🔄 刷新中...", callback=None))
        self._fetch_and_update()

    def _show_settings(self, _):
        response = rumps.Window(
            title="DeepSeek 余额设置",
            message=(
                f"当前 API Key: {self.api_key[:12]}...{self.api_key[-4:] if len(self.api_key) > 16 else ''}\n"
                f"API 地址: {self.base_url}\n"
                f"轮询间隔: {self.config['poll_interval_minutes']} 分钟\n\n"
                "输入新的 API Key（留空不修改）："
            ),
            default_text="",
            ok="保存",
            cancel="取消",
            dimensions=(400, 180)
        ).run()

        if response.clicked and response.text.strip():
            new_key = response.text.strip()
            self.api_key = new_key
            self.config['api_key'] = new_key
            save_config(self.config)
            self._force_refresh(None)
            rumps.notification(
                title="设置已保存",
                subtitle="API Key 已更新",
                message="正在使用新 Key 重新获取余额..."
            )

    def _quit_app(self, _):
        self.timer.stop()
        rumps.quit_application()


# ============================================================
# 入口
# ============================================================
if __name__ == "__main__":
    DeepSeekBalanceApp().run()

File: DeepSeekBalance.swift
#!/usr/bin/env swift

import Cocoa
import Foundation

// MARK: - 模型
struct Config: Codable {
    var api_key: String = ""
    var base_url: String = "https://api.deepseek.com"
    var poll_interval_minutes: Int = 5
}

struct BalanceInfo: Codable {
    let currency: String
    let total_balance: String
    let granted_balance: String
    let topped_up_balance: String
}

struct BalanceResponse: Codable {
    let is_available: Bool
    let balance_infos: [BalanceInfo]
}

struct HistoryRecord: Codable {
    var timestamp: String
    var total: Double
}

// MARK: - Path helpers
let appDir = FileManager.default.currentDirectoryPath
func path(_ name: String) -> String {
    return (appDir as NSString).appendingPathComponent(name)
}

// MARK: - App Delegate
class AppDelegate: NSObject, NSApplicationDelegate {
    var statusItem: NSStatusItem!
    var config = Config()
    var history: [HistoryRecord] = []
    let dateFmt: DateFormatter = {
        let f = DateFormatter(); f.dateFormat = "yyyy-MM-dd HH:mm:ss"; return f
    }()

    func applicationDidFinishLaunching(_ aNotification: Notification) {
        NSApp.setActivationPolicy(.accessory)
        loadConfig()
        loadHistory()

        statusItem = NSStatusBar.system.statusItem(withLength: NSStatusItem.variableLength)
        statusItem.button?.title = "..."

        if let img = NSImage(contentsOfFile: path("icon.png")) {
            img.size = NSSize(width: 18, height: 18)
            statusItem.button?.image = img
            statusItem.button?.imagePosition = .imageLeading
        }

        buildDefaultMenu()
        statusItem.menu = buildDefaultMenu()

        Timer.scheduledTimer(withTimeInterval: TimeInterval(config.poll_interval_minutes * 60), repeats: true) { _ in
            self.fetch()
        }
        fetch()
    }

    // MARK: - Menu
    func buildDefaultMenu() -> NSMenu {
        let m = NSMenu()
        m.autoenablesItems = false
        let r = addItem(m, "🔄 立即刷新", #selector(refresh))
        let s = addItem(m, "⚙️ 设置...", #selector(settings))
        m.addItem(.separator())
        let q = addItem(m, "❌ 退出", #selector(quitApp))
        return m
    }

    func buildFullMenu(_ info: BalanceInfo) -> NSMenu {
        let m = NSMenu()
        m.autoenablesItems = false
        let s = info.currency == "CNY" ? "¥" : "$"
        let t = Float(info.total_balance) ?? 0
        let g = Float(info.granted_balance) ?? 0
        let p = Float(info.topped_up_balance) ?? 0

        addDisabled(m, "📊 总余额: \(s)\(String(format: "%.2f", t))")
        addDisabled(m, "🎁 赠送余额: \(s)\(String(format: "%.2f", g))")
        addDisabled(m, "💳 充值余额: \(s)\(String(format: "%.2f", p))")
        m.addItem(.separator())
        let usage = todayUsage()
        if usage > 0 {
            addDisabled(m, "📉 今日使用: -\(s)\(String(format: "%.2f", usage))")
        }
        m.addItem(.separator())
        addItem(m, "🔄 立即刷新", #selector(refresh))
        addItem(m, "⚙️ 设置...", #selector(settings))
        m.addItem(.separator())
        addItem(m, "❌ 退出", #selector(quitApp))
        return m
    }

    func addDisabled(_ menu: NSMenu, _ title: String) {
        let item = NSMenuItem(title: title, action: nil, keyEquivalent: "")
        item.isEnabled = false
        menu.addItem(item)
    }

    func addItem(_ menu: NSMenu, _ title: String, _ sel: Selector) -> NSMenuItem {
        let item = NSMenuItem(title: title, action: sel, keyEquivalent: "")
        item.target = self
        menu.addItem(item)
        return item
    }

    // MARK: - Config
    func loadConfig() {
        if let d = try? Data(contentsOf: URL(fileURLWithPath: path("config.json"))),
           let c = try? JSONDecoder().decode(Config.self, from: d) {
            config = c
        }
    }

    func saveConfig() {
        if let d = try? JSONEncoder().encode(config) {
            try? d.write(to: URL(fileURLWithPath: path("config.json")))
        }
    }

    // MARK: - History (JSON file)
    func loadHistory() {
        if let d = try? Data(contentsOf: URL(fileURLWithPath: path("data/history.json"))),
           let h = try? JSONDecoder().decode([HistoryRecord].self, from: d) {
            history = h
        }
    }

    func saveHistory() {
        let dir = path("data")
        try? FileManager.default.createDirectory(atPath: dir, withIntermediateDirectories: true)
        if let d = try? JSONEncoder().encode(history) {
            try? d.write(to: URL(fileURLWithPath: path("data/history.json")))
        }
    }

    func todayUsage() -> Float {
        let df = DateFormatter(); df.dateFormat = "yyyy-MM-dd"
        let today = df.string(from: Date())
        let todayRecords = history.filter { $0.timestamp.hasPrefix(today) }
        guard let first = todayRecords.first, let last = todayRecords.last else { return 0 }
        return Float(first.total - last.total)
    }

    // MARK: - API
    func fetch() {
        guard !config.api_key.isEmpty else {
            DispatchQueue.main.async { self.showError("请先设置 API Key") }
            return
        }
        let url = URL(string: "\(config.base_url)/user/balance")!
        var req = URLRequest(url: url, timeoutInterval: 15)
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.setValue("Bearer \(config.api_key)", forHTTPHeaderField: "Authorization")

        URLSession.shared.dataTask(with: req) { data, resp, error in
            if let e = error {
                DispatchQueue.main.async { self.showError(e.localizedDescription) }
                return
            }
            guard let d = data, let http = resp as? HTTPURLResponse else {
                DispatchQueue.main.async { self.showError("无响应") }
                return
            }
            if http.statusCode == 401 {
                DispatchQueue.main.async { self.showError("API Key 无效") }
                return
            }
            guard http.statusCode == 200,
                  let bal = try? JSONDecoder().decode(BalanceResponse.self, from: d),
                  let info = bal.balance_infos.first else {
                DispatchQueue.main.async { self.showError("解析失败") }
                return
            }
            DispatchQueue.main.async { self.update(info: info) }

            // 保存历史
            let rec = HistoryRecord(
                timestamp: self.dateFmt.string(from: Date()),
                total: Double(info.total_balance) ?? 0
            )
            self.history.append(rec)
            if self.history.count > 10000 { self.history.removeFirst(1000) }
            self.saveHistory()
        }.resume()
    }

    func update(info: BalanceInfo) {
        let s = info.currency == "CNY" ? "¥" : "$"
        let t = Float(info.total_balance) ?? 0
        statusItem.button?.title = "\(s)\(String(format: "%.2f", t))"
        statusItem.menu = buildFullMenu(info)
    }

    func showError(_ msg: String) {
        statusItem.button?.title = "⚠️"
        let m = NSMenu()
        m.autoenablesItems = false
        addDisabled(m, "❌ \(msg)")
        m.addItem(.separator())
        addItem(m, "🔄 重试", #selector(refresh))
        addItem(m, "⚙️ 设置...", #selector(settings))
        m.addItem(.separator())
        addItem(m, "❌ 退出", #selector(quitApp))
        statusItem.menu = m
    }

    @objc func refresh() {
        statusItem.button?.title = "..."
        fetch()
    }

    @objc func settings() {
        let alert = NSAlert()
        alert.messageText = "设置 API Key"
        alert.informativeText = "请输入 DeepSeek API Key："
        alert.addButton(withTitle: "保存")
        alert.addButton(withTitle: "取消")

        let tf = NSTextField(frame: NSRect(x: 0, y: 0, width: 300, height: 24))
        tf.placeholderString = "sk-..."
        tf.stringValue = config.api_key
        alert.accessoryView = tf

        if alert.runModal() == .alertFirstButtonReturn {
            let key = tf.stringValue.trimmingCharacters(in: .whitespaces)
            if !key.isEmpty {
                config.api_key = key
                saveConfig()
                refresh()
            }
        }
    }

    @objc func quitApp() {
        NSApp.terminate(nil)
    }
}

let app = NSApplication.shared
let delegate = AppDelegate()
app.delegate = delegate
app.run()
